### PR TITLE
Placeholder types: take two

### PR DESCRIPTION
**Pull request by Jumhyn**
Implements the [placeholder types](https://github.com/apple/swift-evolution/pull/1317) proposal.

### CODE REVIEW

File: lib/Sema/PreCheckExpr.cpp
Comment (Jumhyn) [inline lib/Sema/PreCheckExpr.cpp:1590]: This hack helps us preserve some diagnostics for uses of '`_`' which were probably "supposed" to be patterns/discard assignments, but it's ugly. Would love any suggestions for a better way to handle this!

**Comment by Jumhyn** [comment]
I'm also seeing a crash (that only happens on release builds 😞) in `PlaceholderType::get` whenever we try to resolve the first placeholder type... if anyone sees something obviously suspect here, please let me know!

```
Type PlaceholderType::get(ASTContext &ctx, Originator originator) {
  assert(originator);
  return new (ctx, AllocationArena::Permanent)
      PlaceholderType(ctx, originator, RecursiveTypeProperties::HasPlaceholder);
}
```

File: lib/Sema/PreCheckExpr.cpp
Comment (Jumhyn) [reply lib/Sema/PreCheckExpr.cpp:1590]: Hmm, we could maybe switch from the bottom-up type folding we're currently doing to a top-down approach. 🤔

Comment (Jumhyn) [reply lib/Sema/PreCheckExpr.cpp:1590]: Actually, I think there's a better heuristic we can use here that involves keeping track of information the whole way down the tree.

**Comment by Jumhyn** [comment]
cc @xedin @CodaFi @hborla since y'all looked at the last PR on placeholder types 🙂

**Comment by xedin** [comment]
@Jumhyn This is in my queue, going to try to look asap.

**Review by xedin — COMMENTED** [review]
LGTM! I have left some comments inline, but they are mostly minor. I also think it might be worth it to land some part of this PR separately.

File: lib/Sema/TypeCheckType.h
Comment (Jumhyn) [reply lib/Sema/TypeCheckType.h:279]: Yep I think that makes sense! I will pull this out into a separate PR and tag you on it.

File: lib/Sema/CSApply.cpp
Comment (Jumhyn) [reply lib/Sema/CSApply.cpp:6524]: What exactly are the implications of failing vs. not failing a coercion? Not sure I understand exactly what it means to diagnose an error while failing a coercion vs. not failing a coercion.

File: lib/AST/Expr.cpp
Comment (Jumhyn) [reply lib/AST/Expr.cpp:752]: Yeah that makes total sense to me, especially since the comment insists that it's exhaustive!

File: lib/Sema/PreCheckExpr.cpp
Comment (Jumhyn) [reply lib/Sema/PreCheckExpr.cpp:910]: Yeah I was struggling to come up with clear/concise names for this and the related methods, I'll rework with your suggestions.

File: lib/Sema/TypeCheckType.h
Comment (Jumhyn) [reply lib/Sema/TypeCheckType.h:283]: Opened as #36948. The commit is still in this PR (for the time being) since it's necessary for the changes, but once the other PR is merged I'll make sure it gets dropped from this PR.

File: lib/Sema/CSApply.cpp
Comment (Jumhyn) [reply lib/Sema/CSApply.cpp:6582]: @xedin Relatedly, the docs for `coerceToType` say "This operation cannot fail," but clients are clearly checking against failure of this function... is that comment outdated?

Comment (xedin) [reply lib/Sema/CSApply.cpp:6582]: I think it is. It's possible (albeit rare) for solver to produce a solution which couldn't be applied to AST, in such cases `ExprRewriter` should fail, if coercion is impossible than `coerceToType` should fail as well. In this case `coerceToType` would produce an error diagnostic and should return `nullptr`.

Comment (xedin) [reply lib/Sema/CSApply.cpp:6582]: It it was just a warning then coerceToType should have just continued.

Comment (xedin) [reply lib/Sema/CSApply.cpp:6582]: It might be better to let `diagnoseInvalidFunctionType` return a bool to indicate whether error was diagnosed and only fail `coerceToType` if so.

Comment (Jumhyn) [reply lib/Sema/CSApply.cpp:6582]: Got it, makes sense. Yeah, you're right that to do this we'll have to have `diagnoseInvalidFunctionType` indicate failure.

Looking back at the "old" implementation, it looks like during type resolution we diagnosed these invalid types but didn't fail resolution, so presumably those invalid function types are getting applied to the AST. Do you think it makes sense to fail the application in all cases? I.e., remove the `diagnoseInvalidFunctionType` check from `TypeResolver::resolveASTFunctionType` and perform it only in `coerceToType` for all function types (not just those that had a placeholder), and fail coercion if so?

Comment (xedin) [reply lib/Sema/CSApply.cpp:6582]: Frankly I'm not sure exactly why it wasn't made to fail, I'd suggest you try to return `ErrorType` wrapping `fnTy` for cases when `diagnoseInvalidFunctionType` would produce an error in `resolveASTFunctionType` (so that code completion would still have access to underlying type) and return `nullptr` in `coerceToType`.

Comment (Jumhyn) [reply lib/Sema/CSApply.cpp:6582]: Ok! I’ll give that a shot and see if there’s any unexpected fallout.

Comment (xedin) [reply lib/Sema/CSApply.cpp:6582]: 🤞

**Comment by Jumhyn** [comment]
@xedin So it seems like things mostly work out just fine. The only additional failure is in `test/AutoDiff/Sema/differentiable_func_type.swift`, where issues that were previously diagnosed as ambiguities now get more specific error messages because some overloads became unviable (now that their parameter types are `ErrorType`). The full diff of the necessary diagnostic changes is below—if you think this is acceptable I'll push up the changes!

```
diff --git a/test/AutoDiff/Sema/differentiable_func_type.swift b/test/AutoDiff/Sema/differentiable_func_type.swift
index 832a7411460..3493f50e16b 100644
--- a/test/AutoDiff/Sema/differentiable_func_type.swift
+++ b/test/AutoDiff/Sema/differentiable_func_type.swift
@@ -45,8 +45,7 @@ let _: @differentiable(_linear) (Float) -> Float
 func test1<T: Differentiable, U: Differentiable>(_: @differentiable(reverse) (T) -> @differentiable(reverse) (U) -> Float) {}
 // expected-error @+1 {{result type '(U) -> Float' does not conform to 'Differentiable', but the enclosing function type is '@differentiable'}}
 func test2<T: Differentiable, U: Differentiable>(_: @differentiable(reverse) (T) -> (U) -> Float) {}
-// expected-error @+2 {{result type 'Int' does not conform to 'Differentiable', but the enclosing function type is '@differentiable'}}
-// expected-error @+1 {{result type '@differentiable(reverse) (U) -> Int' does not conform to 'Differentiable', but the enclosing function type is '@differentiable'}}
+// expected-error @+1 {{result type 'Int' does not conform to 'Differentiable', but the enclosing function type is '@differentiable'}}
 func test3<T: Differentiable, U: Differentiable>(_: @differentiable(reverse) (T) -> @differentiable(reverse) (U) -> Int) {}
 // expected-error @+1 {{result type '(U) -> Int' does not conform to 'Differentiable', but the enclosing function type is '@differentiable'}}
 func test4<T: Differentiable, U: Differentiable>(_: @differentiable(reverse) (T) -> (U) -> Int) {}
@@ -189,7 +188,6 @@ extension Vector: Differentiable where T: Differentiable {
 // expected-note@+1 2 {{found this candidate}}
 func inferredConformancesGeneric<T, U>(_: @differentiable(reverse) (Vector<T>) -> Vector<U>) {}
 
-// expected-note  @+5 2 {{found this candidate}}
 // expected-error @+4 {{generic signature requires types 'Vector<T>' and 'Vector<T>.TangentVector' to be the same}}
 // expected-error @+3 {{generic signature requires types 'Vector<U>' and 'Vector<U>.TangentVector' to be the same}}
 // expected-error @+2 {{parameter type 'Vector<T>' does not conform to 'Differentiable' and satisfy 'Vector<T> == Vector<T>.TangentVector', but the enclosing function type is '@differentiable(_linear)'}}
@@ -203,7 +201,8 @@ func nondiff(x: Vector<Int>) -> Vector<Int> {}
 
 // expected-error @+1 {{no exact matches in call to global function 'inferredConformancesGeneric'}}
 inferredConformancesGeneric(nondiff)
-// expected-error @+1 {{no exact matches in call to global function 'inferredConformancesGenericLinear'}}
+// expected-error @+2 {{global function 'inferredConformancesGenericLinear' requires that 'Int' conform to 'Differentiable'}}
+// expected-error @+1 {{global function 'inferredConformancesGenericLinear' requires that 'Int' conform to 'Differentiable'}}
 inferredConformancesGenericLinear(nondiff)
 
 func diff(x: Vector<Float>) -> Vector<Float> {}
@@ -228,13 +227,15 @@ extension Linear: Differentiable where T: Differentiable, T == T.TangentVector {
 // expected-note @+1 2 {{found this candidate}}
 func inferredConformancesGeneric<T, U>(_: @differentiable(reverse) (Linear<T>) -> Linear<U>) {}
 
-// expected-note @+1 2 {{found this candidate}}
+// expected-note  @+2 2 {{where 'T' = 'Int'}}
+// expected-note  @+1 2 {{where 'U' = 'Int'}}
 func inferredConformancesGenericLinear<T, U>(_: @differentiable(_linear) (Linear<T>) -> Linear<U>) {}
 
 func nondiff(x: Linear<Int>) -> Linear<Int> {}
 // expected-error @+1 {{no exact matches in call to global function 'inferredConformancesGeneric'}}
 inferredConformancesGeneric(nondiff)
-// expected-error @+1 {{no exact matches in call to global function 'inferredConformancesGenericLinear'}}
+// expected-error @+2 {{global function 'inferredConformancesGenericLinear' requires that 'Int' conform to 'Differentiable'}}
+// expected-error @+1 {{global function 'inferredConformancesGenericLinear' requires that 'Int' conform to 'Differentiable'}}
 inferredConformancesGenericLinear(nondiff)
 
 func diff(x: Linear<Float>) -> Linear<Float> {}
```

**Comment by xedin** [comment]
@Jumhyn Sorry it took me a bit to respond, I'm on vacation this week. I think new diagnostics make sense, they are definitely better than ambiguities especially considering that other overload is invalid, so I'd say we should proceed with failing in type resolution + failing on `coerceToType`.

**Comment by Jumhyn** [comment]
No worries at all @xedin, hope you're having a nice vacation! Please feel free to ignore me until you're back. :)

For whenever you're back: yes, I agree the diagnostics are improved! Just wanted to make sure I wasn't missing anything. I've rebased to fold the type resolution failure into these changes, so I believe all your comments should be addressed at this point.

**Comment by Jumhyn** [comment]
Hey @xedin, if you're still on vacation then disregard, just wanted to put this back on your radar in case it slipped off while you were out!

**Comment by xedin** [comment]
I'm back and it did slip from my radar unfortunately but now it's back in the fold :)

File: lib/Sema/CSApply.cpp
Comment (xedin) [inline lib/Sema/CSApply.cpp:6583]: Nit: Is this a valid formatting?

**Review by xedin — APPROVED** [review]
LGTM!

File: lib/Sema/CSApply.cpp
Comment (Jumhyn) [reply lib/Sema/CSApply.cpp:6583]: I'll run it thru `clang-format` again just to make sure.

Comment (xedin) [reply lib/Sema/CSApply.cpp:6583]: Thank you!

**Comment by Jumhyn** [comment]
@xedin Alright, updated with `clang-format`'s preferred line breaks for that snippet. Thank you!

**Review by xedin — APPROVED** [review]

**Comment by jckarter** [comment]
Is the [DNM] marker in this title still relevant? Any changes you still want to make before I merge this?

**Comment by Jumhyn** [comment]
@jckarter No, assuming that all tests are still passing (🤞) there aren't any changes I intend to make here. Should be merge-ready!

**Comment by Jumhyn** [comment]
(I can't add labels without commit access, or else would have tagged it as `swift-evolution-pending-discussion`)

**Comment by Jumhyn** [comment]
Thank you @xedin! 😄

**Comment by xedin** [comment]
No problem :)

**Comment by jckarter** [comment]
Thanks! Since this has an implementation now, it should be good to merge and schedule for review.

**Comment by xedin** [comment]
I thought that proposal has to be accepted by the core team before the implementation could be merged... Wouldn’t have had to deal with all the merge conflicts for my features :)

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by jckarter** [comment]
@xedin Oh, sorry, I had both the implementation and evolution PRs open and got them confused. Yes, the implementation should wait to merge until the proposal is accepted.

**Comment by Jumhyn** [comment]
Thanks for clarifying @jckarter!

**Comment by Jumhyn** [comment]
@xedin Looks like test failure is due to returning `ErrorType` from `resolveASTFunctionType` when `diagnoseInvalidFunctionType` fails, the following change fixes it:

```
diff --git a/test/attr/attr_objc.swift b/test/attr/attr_objc.swift
index 6fc79b06edc..9e7dcea642b 100644
--- a/test/attr/attr_objc.swift
+++ b/test/attr/attr_objc.swift
@@ -1380,7 +1380,7 @@ class infer_instanceVar1 {
   var var_CFunctionPointer_1: @convention(c) () -> ()
   // CHECK-LABEL: @objc var var_CFunctionPointer_invalid_1: Int
   var var_CFunctionPointer_invalid_1: @convention(c) Int // expected-error {{@convention attribute only applies to function types}}
-  // CHECK-LABEL: {{^}} var var_CFunctionPointer_invalid_2: @convention(c) (PlainStruct) -> Int
+  // CHECK-LABEL: {{^}} var var_CFunctionPointer_invalid_2: <<error type>>
   var var_CFunctionPointer_invalid_2: @convention(c) (PlainStruct) -> Int // expected-error {{'(PlainStruct) -> Int' is not representable in Objective-C, so it cannot be used with '@convention(c)'}}
   
   // <rdar://problem/20918869> Confusing diagnostic for @convention(c) throws
```

I've pushed the latest up, so if that change seems reasonable to you this is ready for another test run!

**Comment by xedin** [comment]
Since the error has been properly diagnosed I think that's correct behavior.

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by Jumhyn** [comment]
@xedin The macOS failure here at least seems like a flaky test to me. I'm able to reproduce it locally but only intermittently. On my machine it seems like the failure rate is ~50%. Have you seen this flake before?

(Can't say for sure about Linux/Windows since I don't have a setup for running those locally, but since it's the same test/assertion failing it seems likely that it's the same root cause...)

**Comment by xedin** [comment]
Hm, I don't think I have seen this happen in other PRs, might be worth investigating what is going on here...

**Comment by Jumhyn** [comment]
@swift-ci please test

File: lib/Sema/CSDiagnostics.cpp
Comment (CodaFi) [inline lib/Sema/CSDiagnostics.cpp:7319]: Do we have access to any potential bindings here? It would be very cool to surface notes that tell the user what we tried.

Comment (Jumhyn) [reply lib/Sema/CSDiagnostics.cpp:7319]: Good idea! I'll give it a look at see if there's a good way to surface any of that information...

**Comment by Jumhyn** [comment]
@swift-ci please test macOS

**Comment by Jumhyn** [comment]
@swift-ci please build macOS

**Comment by Jumhyn** [comment]
@swift-ci please test macOS

**Comment by Jumhyn** [comment]
@swift-ci please test Windows

**Comment by Jumhyn** [comment]
@swift-ci please test macOS

**Comment by hborla** [comment]
@swift-ci please build toolchain

**Comment by Jumhyn** [comment]
@swift-ci please build toolchain

**Comment by Jumhyn** [comment]
@swift-ci please test macOS

**Comment by hborla** [comment]
macOS CI is not having a good week! I think the failing tests are disabled now.

@swift-ci please test macOS platform

**Comment by hborla** [comment]
@swift-ci please build toolchain macOS platform

**Comment by Jumhyn** [comment]
Thanks @hborla!

File: lib/Sema/CSDiagnostics.cpp
Comment (hborla) [reply lib/Sema/CSDiagnostics.cpp:7319]: Hmm, I would think that if a non-hole potential binding was attempted and didn't work, a more specific fix would be recorded. If several different bindings were attempted and none of them worked but the solutions all had the same score, we need to diagnose that through the ambiguity mechanism, similar to the "conflicting arguments to generic parameter" message. Here are some different examples with generic arguments, and I think placeholders should behave similarly:

```swift
protocol P {}

struct Generic<T>  {
  init() {}
  init(arg: T) where T: P {}
  init(arg1: T, arg2: T) {}
}

func test() {
  // no bindings for T -> missing generic arg fix 
  let generic1: Generic = .init() // error: Generic parameter 'T' could not be inferred

  // one binding for T that didn't work -> specific fix 
  let generic2: Generic = .init(arg: 1) // error: Initializer 'init(arg:)' requires that 'Int' conform to 'P'

  // several bindings for T and none worked -> ambiguity
  let generic3: Generic = .init(arg1: 1, arg2: "") // error: Conflicting arguments to generic parameter 'T' ('Int' vs. 'String')
}
```

One thing that would be cool, but is probably out of scope for this initial implementation, is recording any requirements on missing placeholders so the user has more information about what it needs to be filled in with.

File: lib/Sema/TypeCheckType.cpp
Comment (hborla) [inline lib/Sema/TypeCheckType.cpp:1981]: Should this come with a fix-it to remove the `_`? Or perhaps a note+fix-it saying something along the lines of `remove '_' to infer the top-level type?`?

Comment (Jumhyn) [reply lib/Sema/TypeCheckType.cpp:1981]: Hmm, yeah, I think that makes sense in most cases. We'd need to remove more than just the underscore (e.g. "`: _`" in `let x: _ = 0` or "` as _`" in `0 as _`), but some situations are a bit more complex—e.g. we probably want to turn `_()` into `.init()` since removing `_` produces nonsense. In any case we'd have to move this check outside of the type resolver to a place where we have a bit more context... 🤔

Comment (hborla) [reply lib/Sema/TypeCheckType.cpp:1981]: Ahh good point. I think the fix-it is less important than making the message more actionable, but I'm not quite sure how to word it

File: lib/Sema/TypeCheckType.cpp
Comment (rintaro) [inline lib/Sema/TypeCheckType.cpp:1978]: Does this allow `(_).self`? I think `TyR` is `TupleTypeRepr` at this point.

Comment (Jumhyn) [reply lib/Sema/TypeCheckType.cpp:1978]: Ha! Yeah, I think you're right, good catch.

**Comment by Jumhyn** [comment]
@swift-ci please test macOS

**Comment by Jumhyn** [comment]
@swift-ci please build toolchain

**Comment by Jumhyn** [comment]
@swift-ci please test Linux

**Comment by Jumhyn** [comment]
@swift-ci please test source compatibility

**Comment by Jumhyn** [comment]
@swift-ci please build toolchain macOS

**Comment by Jumhyn** [comment]
@swift-ci please test

**Comment by xedin** [comment]
Hi @Jumhyn! @hborla mentioned to me today that the proposal for these change has been accepted (SE-0315), would you be able to rebase the changes so we can get CI going and landed it? Sorry we have dropped the ball on this.

**Comment by Jumhyn** [comment]
@xedin Not at all, I’ve been distracted as well. I’ve been meaning to get back to this but was on vacation last week. There’s still a bit of cleanup to do here after taking Rintaro’s suggestion from above—hoping to have a chance this week to get that done and will ping you when it’s ready!

Thanks both for following up on this.

**Comment by xedin** [comment]
Sounds good!

**Comment by Jumhyn** [comment]
@swift-ci please smoke test

File: lib/Sema/TypeCheckType.cpp
Comment (Jumhyn) [inline lib/Sema/TypeCheckType.cpp:1942]: @rintaro Updated this portion to check the canonical type! Turned out to be a bit more annoying than expected since a `PlaceholderTypeRepr` isn't always transformed into a `PlaceholderType`—the transformation is based on a client-specified function.

Now, we resolve every `PlaceholderTypeRepr` into a `PlaceholderType`, then after diagnosing top-level placeholders we proceed with the transformation step below. Since the logic here has changed nontrivially would love if you could give it another look!

**Comment by Jumhyn** [comment]
@swift-ci please test

**Comment by Jumhyn** [comment]
@swift-ci please test source compatibility

**Comment by Jumhyn** [comment]
@swift-ci build toolchain macOS

**Comment by Jumhyn** [comment]
@swift-ci please test source compatibility

**Comment by Jumhyn** [comment]
@swift-ci please smoke test

**Comment by Jumhyn** [comment]
@swift-ci please test source compatibility release

**Comment by Jumhyn** [comment]
@xedin This should be ready for another look at this point. It looks like something unrelated is still plaguing the source compat tests, but the main tests are passing now so I don't _think_ any more significant changes will be required. (And a previous source compat run passed everything that wasn't broken before Holly merged #38929)

**Comment by hborla** [comment]
I think the source compatibility failure will be fixed by https://github.com/apple/swift/pull/38928, which was just merged.

@swift-ci please test source compatibility

**Comment by Jumhyn** [comment]
Thank you @hborla! @xedin, looks like this is good to go!

**Review by xedin — APPROVED** [review]

**Comment by xedin** [comment]
@hborla, @rintaro please take a look and we'll land this if everything looks good.

File: lib/Parse/ParseType.cpp
Comment (rintaro) [inline lib/Parse/ParseType.cpp:1477]: (Super nit-pick)
```suggestion
    consumeToken();
```

File: lib/Sema/TypeCheckType.cpp
Comment (rintaro) [inline lib/Sema/TypeCheckType.cpp:1958]: Checking `getPlaceholderHandler()` first looks more efficient.
```swift
  if (const auto handlerFn = resolution->getPlaceholderHandler()) {
    result = result.get().transform.transform([&](Type ty) {
       ...
    });
  }
  return result;
```

**Review by rintaro — APPROVED** [review]
LGTM. Just a couple of nitpicks

File: lib/Sema/TypeCheckType.cpp
Comment (Jumhyn) [reply lib/Sema/TypeCheckType.cpp:1958]: Ah yeah, this should be stable over a given resolution. Good call!

**Comment by Jumhyn** [comment]
Thanks @rintaro, your comments have been addressed!

**Review by rintaro — APPROVED** [review]

**Comment by xedin** [comment]
@swift-ci please test

**Review by hborla — APPROVED** [review]

**Comment by Jumhyn** [comment]
@swift-ci please test

**Comment by Jumhyn** [comment]
@xedin Good to merge?

**Comment by rintaro** [comment]
:tada: @Jumhyn Could you update CHANGELOG.md too?

**Comment by Jumhyn** [comment]
@rintaro Already opened the PR [here](https://github.com/apple/swift/pull/38976), just waiting for tests to complete after the CI outage :)

**Comment by philipturner** [comment]
> @xedin So it seems like things mostly work out just fine. The only additional failure is in `test/AutoDiff/Sema/differentiable_func_type.swift`, where issues that were previously diagnosed as ambiguities now get more specific error messages because some overloads became unviable (now that their parameter types are `ErrorType`). Also, we no longer diagnose two layers deep on a non-differentiable function with a non-differentiable function result—we'll only diagnose the innermost issue, looks like. The full diff of the necessary diagnostic changes is below—if you think this is acceptable I'll push up the changes!
> 
> ```
> -// expected-error @+2 {{result type 'Int' does not conform to 'Differentiable', but the enclosing function type is '@differentiable'}}
> -// expected-error @+1 {{result type '@differentiable(reverse) (U) -> Int' does not conform to 'Differentiable', but the enclosing function type is '@differentiable'}}
> +// expected-error @+1 {{result type 'Int' does not conform to 'Differentiable', but the enclosing function type is '@differentiable'}}
>  func test3<T: Differentiable, U: Differentiable>(_: @differentiable(reverse) (T) -> @differentiable(reverse) (U) -> Int) {}
> ...
> -// expected-note  @+5 2 {{found this candidate}}
> ...
> -// expected-error @+1 {{no exact matches in call to global function 'inferredConformancesGenericLinear'}}
> +// expected-error @+2 {{global function 'inferredConformancesGenericLinear' requires that 'Int' conform to 'Differentiable'}}
> +// expected-error @+1 {{global function 'inferredConformancesGenericLinear' requires that 'Int' conform to 'Differentiable'}}
>  inferredConformancesGenericLinear(nondiff)
> 
> -// expected-note @+1 2 {{found this candidate}}
> +// expected-note  @+2 2 {{where 'T' = 'Int'}}
> +// expected-note  @+1 2 {{where 'U' = 'Int'}}
>  func inferredConformancesGenericLinear<T, U>(_: @differentiable(_linear) (Linear<T>) -> Linear<U>) {}
> ...
> -// expected-error @+1 {{no exact matches in call to global function 'inferredConformancesGenericLinear'}}
> +// expected-error @+2 {{global function 'inferredConformancesGenericLinear' requires that 'Int' conform to 'Differentiable'}}
> +// expected-error @+1 {{global function 'inferredConformancesGenericLinear' requires that 'Int' conform to 'Differentiable'}}
>  inferredConformancesGenericLinear(nondiff)
> ```
> 
> Note: we get the diagnostic twice because the parameter and result type both require that `Int` conform to `Differentiable`.

@Jumhyn this is very likely the source of SR-15808.

**Pull request merged**
